packages/quantic/force-app/main/default/lwc/headlessLoader/headlessLoader.js

```diff
@@ -0,0 +1,145 @@
+    const later = () => {
+      clearTimeout(timeout);
+      func(...args);
+    };
+    clearTimeout(timeout);
+    // eslint-disable-next-line @lwc/lwc/no-async-operation
+    timeout = setTimeout(later, wait);
+  };
+};
+
+/**
+ * Cancels the delayed search query.
+ */
+const cancelInitialSearch = () => {
+  if (timeout) {
+    clearTimeout(timeout);
+    timeout = undefined;
+  }
+}
+
+/**
+ * Dispatches search request.
+ */
+const executeInitialSearch = debounce(() => {
+  window.coveoHeadless.engine.dispatch(
+    CoveoHeadless.SearchActions.executeSearch(
+      CoveoHeadless.AnalyticsActions.logInterfaceLoad()
+    )
+  );
+}, DEBOUNCE_DELAY);
+
+/**
+ * Returns true if registered components are initialized, false otherwise.
+ */
+const getAreComponentsReady = () => !window.coveoHeadless.components.find(component => component.initialized === false);
+
+/**
+ * Loads dependencies and returns an initialized Headless engine. 
+ * @param element 
+ */
+async function initEngine(element) {
+  let engine;
+  try {
+    await loadScript(element, HeadlessPath + '/browser/headless.js');
+    await loadScript(element, AtomicPath + '/atomic-utils.js');
+  
+    const config = CoveoHeadless.HeadlessEngine.getSampleConfiguration();
+  
+    engine = new CoveoHeadless.HeadlessEngine({
+      configuration: config,
+      reducers: CoveoHeadless.searchAppReducers,
+    });
+  } catch (error) {
+    console.error('Fatal error: unable to initialize Coveo Headless', error);
+  }
+  return engine;
+}
+
+/**
+ * Registers a component for future initialization.
+ * @param element 
+ */
+function registerComponentForInit(element) {
+  cancelInitialSearch();
+  if (!window.coveoHeadless) {
+    window.coveoHeadless = {
+      components: [],
+      engine: undefined
+    }
+  }
+  const isComponentRegistered = window.coveoHeadless.components.find((component) => component.element === element)
+  if (!isComponentRegistered) {
+    window.coveoHeadless.components.push({
+      element,
+      initialized: false
+    });
+  }
+}
+
+/**
+ * Sets registered component to initialized.
+ * @param element 
+ */
+function setComponentInitialized(element) {
+  const component = window.coveoHeadless ? window.coveoHeadless.components.find((comp) => comp.element === element) : undefined;
+  if (!component) {
+    throw new Error('Fatal Error: Component was not registered before initialization');
+  }
+  component.initialized = true;
+  if (getAreComponentsReady()) {
+    executeInitialSearch();
+  }
+}
+
+/**
+ * Returns headless engine promise.
+ * @param element The Lightning Element component with which to load dependencies.
+ */
+function getHeadlessEngine(element) {
+  if (window.coveoHeadless.engine) {
+    return Promise.resolve(window.coveoHeadless.engine);
+  } 
+  window.coveoHeadless.engine = initEngine(element);
+  window.coveoHeadless.engine.then((engine) => {
+    window.coveoHeadless.engine = engine;
+  });
+  return window.coveoHeadless.engine;
+}
```

Comment:
Is it intended that `window.coveoHeadless.engine` is polymorphic ? (`CoveoHeadless.HeadlessEngine | Promise<CoveoHeadless.HeadlessEngine>`)

Reply:
Yes but there might be a better way to do it. What i want is for this function to return a promise that resolves with an engine instance as its value, even if the original promise has resolved already.

Reply:
If so then L119-L121 is useless.
When you have a promise that is resolve, it state won't change.
Thus, any code that will do `window.coveoHeadless.engine.then` will have access to the engine even if it's done many times.

Is-it clear enough ? 🙂

Reply:
That's what i thought but I added those lines in response to an error. It was several commits ago so i'll reinvestigate

Reply:
Yep i'll remove those lines. the original error was cause by a misuse on my part. 👍